bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java

```diff
@@ -243,15 +246,30 @@ public String toString() {
      */
     @Override
     public void stateChanged(Item item, State oldState, State newState) {
-        setState(function.calculate(members));
+        setStateWithouEvent(function.calculate(members));
+        sendGroupStateChangedEvent(item.getName(), this.state, oldGroupState);
     }
-
+    
     /**
      * @{inheritDoc
      */
     @Override
     public void stateUpdated(Item item, State state) {
-        setState(function.calculate(members));
+        setStateWithouEvent(function.calculate(members));
+        sendGroupStateChangedEvent(item.getName(), this.state, oldGroupState);
+    }
+    
+    private void setStateWithouEvent(State state) {
+        this.oldGroupState = this.state;
+        this.state = state;
+        notifyListeners(oldGroupState, state);
     }
 
+    private void sendGroupStateChangedEvent(String memberName, State newState, State oldState) {
+        if (eventPublisher != null && !oldGroupState.equals(this.state)) {

```

Comment:
Can't you use the oldState instead of the oldGroupState and newState instead of this.state here? Would be much nicer to refer to the passed parameters than to instance variables.